src/nbl/ext/MitsubaLoader/CMitsubaMaterialCompilerFrontend.cpp

```diff
@@ -86,27 +87,99 @@ CMitsubaMaterialCompilerFrontend::EmitterNode* CMitsubaMaterialCompilerFrontend:
-            profile.up = core::normalize(inverseTransform[1]);
-            profile.view = core::normalize(inverseTransform[2]);
 
-            res->emissionProfile = profile;
-            float normalizeEnergy = _emitter->area.emissionProfile->normalizeEnergy;
-            if (normalizeEnergy == 0.0f) {
-                res->intensity *= meta->getMaxIntensity();
+            const bool inFlattenDomain = _emitter->area.emissionProfile->flatten >= 0.0 && _emitter->area.emissionProfile->flatten < 1.0; // [0, 1)
+
+            if (!inFlattenDomain)
+            {
+                const auto cFlatten = core::min(core::abs(_emitter->area.emissionProfile->flatten), 1.f - std::numeric_limits<float>::epsilon());
+                os::Printer::log("WARNING: Flatten property = " + std::to_string(_emitter->area.emissionProfile->flatten) + " is outside it's [0, 1) domain - the flatten property will be clamped to " + std::to_string(cFlatten), ELL_WARNING);
+                _emitter->area.emissionProfile->flatten = cFlatten;
             }
-            else if (normalizeEnergy > 0.0f) {
-                res->intensity *= normalizeEnergy * meta->getMaxIntensity() / meta->getIntegral();
+
+            if (_emitter->area.emissionProfile->flatten > 0.0) // if 0 then it's just "image"
+            {
+                core::smart_refctd_ptr<asset::ICPUImageView> flattenIES = nullptr;
+
+                const auto aHash = _emitter->area.emissionProfile->filename + "?flatten=" + std::to_string(_emitter->area.emissionProfile->flatten);
+                const asset::IAsset::E_TYPE types[]{ asset::IAsset::ET_IMAGE_VIEW, asset::IAsset::ET_TERMINATING_ZERO };
+                asset::SAssetBundle bundle = m_loaderContext->override_->findCachedAsset(aHash, types, m_loaderContext->inner, 0u);
+                {
+                    auto contents = bundle.getContents();
+
+                    if (!contents.empty() && bundle.getAssetType() == asset::IAsset::ET_IMAGE_VIEW)
+                        flattenIES = core::smart_refctd_ptr_static_cast<asset::ICPUImageView>(*contents.begin());
+                }
+
+                if (!flattenIES)
+                {
+                    flattenIES = meta->profile.createIESTexture(_emitter->area.emissionProfile->flatten);
+
+                    if (flattenIES)
+                    {
+                        asset::IAssetLoader::SAssetLoadContext ctx = { {}, nullptr };
+                        asset::SAssetBundle bundle = asset::SAssetBundle(nullptr, { core::smart_refctd_ptr(flattenIES) });
+                        m_loaderContext->override_->insertAssetIntoCache(bundle, aHash, m_loaderContext->inner, 0u);
+                    }
+                    else
+                    {
+                        os::Printer::log("WARNING: Failed to flatten IES \"" + _emitter->area.emissionProfile->filename + "\" texture, using original one", ELL_WARNING);
+                        flattenIES = core::smart_refctd_ptr(image);
+                    }
```

Comment:
else no IES profile at all, because we'll load metadata only on the `cacheEmissionProfile`

Reply:
resolved